### PR TITLE
smallbaselineApp_hyp3 for a simple HyP3 + MintPy workflow

### PR DESCRIPTION
**Pull request by cirrusasf**
…rk with hyp3 product

### CODE REVIEW

**Comment by yunjunz** [comment]
This is great @cirrusasf!

I would recommend naming it as `prep_TS_hyp3` for shorter names.

As for the contents, I would expect some changes once we cleared and merged https://github.com/insarlab/MintPy/pull/581. So let's wait for that.

File: Introduction_hyp3_mintpy.ipynb [hunk 1]
```diff
@@ -0,0 +1,500 @@
...
+    "Before running the program, some package need to be installed and imported."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "!conda install -y -c conda-forge hyp3_sdk"
```
Comment (asjohnston-asf) [inline Introduction_hyp3_mintpy.ipynb:39]: I'd recommend installing `ipywidgets` here as well.  `hyp3_sdk.watch()` requires `ipywidgets` in a jupyter environment, but it's not included by default, leading to an error in the `watch()` cell later in the notebook.

Screenshot from 2021-06-16 13-58-53

File: prep_TS_hyp3.ipynb [hunk 2]
```diff
@@ -0,0 +1,500 @@
...
+   "outputs": [],
+   "source": [
+    "# get files to be clipped in the outdir and its sub-directories.\n",
+    "files = []\n",
+    "pattern = [\"*.tif\"]                      \n",
+    "for dir,_,_ in os.walk(outdir):\n",
+    "    for item in pattern:\n",
+    "        files.extend(glob.glob(os.path.join(dir, item)))\n",
+    "# clip off files\n",
+    "cutGeotiffsByLine(files)    "
```
Comment (asjohnston-asf) [inline prep_TS_hyp3.ipynb:306]: Generating the list of tif files can be simplified using wildcards:
```
pattern = os.path.join(outdir, '**/*.tif')
files = glob.glob(pattern)
cutGeotiffsByLine(files)
```

File: prep_TS_hyp3.ipynb [hunk 3]
```diff
@@ -0,0 +1,500 @@
...
+    "\n",
+    "config = f'''\n",
+    "mintpy.load.processor        = hyp3\n",
+    "##---------interferogram datasets:\n",
+    "mintpy.load.unwFile          = {outdir}/*/*unw_phase_clip.tif\n",
+    "mintpy.load.corFile          = {outdir}/*/*corr_clip.tif\n",
+    "##---------geometry datasets:\n",
+    "mintpy.load.demFile          = {outdir}/DEM/dem_clip.tif\n",
+    "##---------incAngleFile:\n",
+    "mintpy.load.incAngleFile    =  {outdir}/*/*inc_map_clip.tif\n",
```
Comment (asjohnston-asf) [inline prep_TS_hyp3.ipynb:390]: Testing suggests we can use wildcards for both the dem and the inc_map, and *one* of the matching files will be chosen and used for processing.  That should be quick and easy for most use cases, and should eliminate the need for the earlier `Copy metadata file and dem file to DEM subdirectory`
```suggestion
    "##---------geometry datasets:\n",
    "mintpy.load.demFile          = {outdir}/*/*dem_clip.tif\n",
    "mintpy.load.incAngleFile     =  {outdir}/*/*inc_map_clip.tif\n",
```

Comment (cirrusasf) [reply prep_TS_hyp3.ipynb:390]: I suggest keep as it is. The smallbaselineApp loads the dem_clip.tif and the metadata file which is called dem_clip.txt.  If follow your suggestion, we have to rename every *.txt file into dem_clip.txt.  I think current config file make a clear idea how we get the dem.tif and its metadata.txt.

File: prep_TS_hyp3.ipynb (hunk 2 above)
Comment (cirrusasf) [reply prep_TS_hyp3.ipynb:306]: accept your suggestion

File: prep_TS_hyp3.ipynb (hunk 3 above)
Comment (yunjunz) [reply prep_TS_hyp3.ipynb]: I second @asjohnston-asf suggestion. With the latest version from Forrest, it works from my testing. Plus, no need for manual DEM dir.

Comment (cirrusasf) [reply prep_TS_hyp3.ipynb]: changed as your suggestion

**Comment by asjohnston-asf** [comment]
Because HyP3 InSAR geotiffs should always be aligned to a common pixel grid, I believe a simple gdal.Translate call using the `projWin` option is sufficient to clip all the input geotiffs to a common width and height.  I think the "Clip the geotiff files in the current directory and its subdirectories" section can be reduced to:
```
from osgeo import gdal


pattern = os.path.join(outdir, '**/*.tif')
files = glob.glob(pattern)

# determine the smallest area covered by all input files
corners = [gdal.Info(f, format='json')['cornerCoordinates'] for f in files]
ulx = max(corner['upperLeft'][0] for corner in corners)
uly = min(corner['upperLeft'][1] for corner in corners)
lrx = min(corner['lowerRight'][0] for corner in corners)
lry = max(corner['lowerRight'][1] for corner in corners)

# subset all input files to these common coordinates
for f in files:
    output_filename = f.replace('.tif', '_clip.tif')
    gdal.Translate(destName=output_filename, srcDS=f, projWin=[ulx, uly, lrx, lry])
```

**Comment by yunjunz** [comment]
The current workflow is generating the nearest neighbor interferometric pairs. For small baseline approach, more redundant **small baseline** pairs are desired. I hope you guys to add this piece of functionality here. One example from mintpy: https://github.com/insarlab/MintPy/blob/main/mintpy/utils/network.py#L583.

**Comment by cirrusasf** [comment]
Hi YunJun,

As the example in the notebook, we produce the time-series analysis for the
2019 Ridgecrest earthquakes. Attached are the result figures. Would you
please take a look at them and make sure the example is good for the
notebook. After you confirm,  we will hardcord the example in the notebook,
and modify the dir_structure if necessary.

Thank you,

Jiang

**Comment by yunjunz** [comment]
Hi @cirrusasf, please pay attention to the modifications I have made on your PR, otherwise, it would be a waste of time to go back and forth.

It seems like the majority of this notebook is about InSAR processing (search, job submission, etc.). The data loading and running part of mintpy is simple and could be easily covered in the [doc of dir structure](https://mintpy.readthedocs.io/en/latest/dir_structure/#asf_hyp3). Thus, it seems to me that hosting this notebook on the ASF repo is more suitable and convenient for development and maintenance in the long term. If so, we could just add a link in this repo for it, as we did for the [ARIA notebook](https://github.com/insarlab/MintPy-tutorial#3-manipulate-aria-standard-insar-products-with-aria-tools-video). Let me know please.

**Comment by jhkennedy** [comment]
@yunjunz you right a lot of this notebook is more about submitting HyP3 jobs. We were trying to eliminate ping-ponging users across HyP3 and MintPy documentation as much as possible and provide a self-contained walkthrough. 

If we host the full tutorial (which we could absolutely do), I have concerns about us:
1. having to setup a MintPy environment for users or explain how to do so
2. ensuring that the environment/explanation and any MintyPy code is up-to date as MintPy develops.

I think (1) and (2) could be mostly addressed by providing MintPy as an installable package on conda-forge so we can pin versions and watch for releases (we could help you with that!) 

One option would be to:
* Update the HyP3 InSAR example to use the 2019 Ridgecrest earthquake 
* Start this notebook with the assumption that users have downloaded that example into a directory (eliminating sections 1 and 2 of this notebook

If you think pointing to our example and Sections 3+4 of this notebook could be adequately addressed in docs of the dir structure, that'd work too.

**Comment by yunjunz** [comment]
@jhkennedy it would certainly ease the env setup/installation if you could help to make `conda install mintpy` work and you are more than welcomed to do so.

Regarding the HyP3 + MintPy example workflow, I like the idea of separating sections 1, 2 and 3 from section 4 as two independent notebooks. The assumption of an already downloaded directory can be shown using a prepared example, such as the [San Francisco one for ARIA](https://zenodo.org/record/4320005). That would serve as a convention between HyP3 and MintPy as well.

I still think hosting the notebook for sections 1-3 in ASF is better because I expect these parts would evolve and get even more sophisticated and/or convenient in the future, e.g. easier auto grabbing of all the SLC filenames using (SNWE, start/end date, track number, asc/desc, etc.), more sophisticated pairs selection, run section 3 in one line with a function or script, etc. It makes more sense for your guys to develop and maintain it where you have full permission and control.

And the notebook for section 4 is essentially the same in mintpy for different InSAR processors, and there are already two notebooks here for it with very detailed explanations. Hosting this notebook in either mintpy-tutorial or ASF sounds good to me.

**Comment by cirrusasf** [comment]
Hi Yunjun,

I have a question to ask your opinion. We are going to deliver the
water-mask geotiff file. We use 0 to present no-valid data (ocean), As far
as valid data is concerned, we can use either 1 or 255 to represent it. Do
you have an opinion on this?

Thank you,

Jiang

**Comment by yunjunz** [comment]
Hi @cirrusasf, either of them would work, and I would think 1 is better as it's more consistent with other InSAR products as provided by ISCE and ARIA.

**Comment by cirrusasf** [comment]
Hi Yunjun,

Thank you for your clarification.

Jiang

**Review by yunjunz — APPROVED** [review]
Looks good to me. Thank you @cirrusasf.

For anyone who might be interested: the notebook can be easily reviewed here: https://app.reviewnb.com/insarlab/MintPy-tutorial/pull/22/

**Comment by yunjunz** [comment]
@cirrusasf could you resolve the conflicts flagged by GitHub so that we could merge this PR?

**Comment by cirrusasf** [comment]
I have no idea what I need to do for your request:

@cirrusasf could you resolve the conflicts flagged by GitHub so that we could merge this PR?

Here is a reference: https://github.com/insarlab/MintPy/blob/main/docs/CONTRIBUTING.md#divergence-between-upstreammain-and-your-feature-branch

**Comment by cirrusasf** [comment]
Thank you very much!

**Comment by cirrusasf** [comment]
Hi  Yunjun,

We released the new version of the hyp3. We include the water_mask_map.tif
in the new release.

Thank you,

Jiang

**Comment by yunjunz** [comment]
Great, thank you for the notification.

May I ask what's the status of the incidence angle? The non-local version that mintpy expects. We should update the corresponding documents and notebooks after this is done.

**Comment by cirrusasf** [comment]
Hi Yunjun,

We are about to release this.  I think by the end of next week. I will let
you know when it is released.

Jiang

**Comment by cirrusasf** [comment]
Hi Yunjun,

Would you please test if the water_mask works in the MintPy and let us know
the result?

Thank you very much,

Jiang

**Pull request merged**